test/unit/providers/simulation-provider.test.ts

```diff
@@ -4,108 +4,206 @@ import { Percent } from '@uniswap/sdk-core';
-      fromAddressMock,
-      swapOptionsMock,
-      swapRouteMock,
-      amountMock,
-      quoteMock
+      fromAddress,
+      swapOptions,
+      swaproute,
+      amount,
+      quote
     );
-    expect(simulateTxStub.calledOnce).toBeTruthy();
+
+    expect(ethEstimateGasSimulator.ethEstimateGas.calledOnce).toBeTruthy();
     expect(swapRoute.simulationStatus).toEqual(SimulationStatus.Succeeded);
   });
 
-  test('does not simulate when user does not have sufficient balance', async () => {
-    sinon.replace(
+  test('simulates through tenderly when user has sufficient balance but not approval', async () => {
+    sinon.stub(simulator, <any>'userHasSufficientBalance').resolves(true);
+    sinon.stub(
       simulator,
-      <any>'userHasSufficientBalance',
-      async () => false
+      <any>'checkTokenApproved',
+    ).resolves(false);
+    tenderlySimulator.simulateTransaction.resolves({simulationStatus: SimulationStatus.Succeeded} as SwapRoute);
+    const swapRoute = await simulator.simulate(
+      fromAddress,
+      swapOptions,
+      swaproute,
+      amount,
+      quote
     );
+
+    expect(ethEstimateGasSimulator.ethEstimateGas.called).toBeFalsy();
+    expect(tenderlySimulator.simulateTransaction.calledOnce).toBeTruthy();
+    expect(swapRoute.simulationStatus).toEqual(SimulationStatus.Succeeded);
+  });
+
+  test('does not simulate when user does not have sufficient balance', async () => {
+    sinon.stub(
+      simulator,
+      <any>'userHasSufficientBalance'
+    ).resolves(false);
     const swapRoute = await simulator.simulate(
-      fromAddressMock,
-      swapOptionsMock,
-      swapRouteMock,
-      amountMock,
-      quoteMock
+      fromAddress,
+      swapOptions,
+      swaproute,
+      amount,
+      quote
     );
-    expect(simulateTxStub.called).toBeFalsy();
-    expect(swapRoute.simulationStatus).toEqual(SimulationStatus.Unattempted);
+    expect(ethEstimateGasSimulator.ethEstimateGas.called).toBeFalsy();
+    expect(tenderlySimulator.simulateTransaction.called).toBeFalsy();
+    expect(swapRoute.simulationStatus).toEqual(SimulationStatus.InsufficientBalance);
   });
 });
+describe('Eth estimate gas simulator', () => {
+    const chainId = ChainId.MAINNET;
+    const simulator = new EthEstimateGasSimulator(chainId, provider, v2PoolProvider, v3PoolProvider);
+    let ethEstimateGasStub: sinon.SinonStub;
+
+    const swaproute: SwapRoute = {
+        quote: quote,
+        quoteGasAdjusted: quote,
+        estimatedGasUsed: estimatedGasUsed,
+        estimatedGasUsedQuoteToken: quote,
+        estimatedGasUsedUSD: quote,
+        gasPriceWei: estimatedGasUsed,
+        trade: trade as Trade<any, any, any>,
+        route: route,
+        blockNumber: blockNumber,
+        simulationStatus: SimulationStatus.Succeeded,
+    };
+
+    beforeEach(() => {
+        ethEstimateGasStub = sinon.stub(simulator, <any>'ethEstimateGas');
+        ethEstimateGasStub.resolves(swaproute)
+    });
```

Comment:
In these unit tests you are mocking functions of the class, so the result is we aren't testing all the logic in the class. i.e. all the logic in `ethEstimateGas` about `swap_router_02` vs `universal_router`, gas multiplying, setting status to failed or not, is all untested.

The correct way to do it is to mock the dependencies that are injected into the class. So you should be mocking the jsonRpcProvider to return different things

Reply:
I see, doing that now. And will keep in mind for the future.